reporter/block_handler.go

```diff
@@ -2,35 +2,59 @@ package reporter
 
 import (
 	"fmt"
-
 	"github.com/babylonlabs-io/vigilante/types"
+	"github.com/btcsuite/btcd/wire"
 )
 
 // blockEventHandler handles connected and disconnected blocks from the BTC client.
 func (r *Reporter) blockEventHandler() {
 	defer r.wg.Done()
 	quit := r.quitChan()
 
+	if err := r.btcNotifier.Start(); err != nil {
+		r.logger.Errorf("Failed starting notifier")
+		return
+	}
+
+	blockNotifier, err := r.btcNotifier.RegisterBlockEpochNtfn(nil)
+	if err != nil {
+		r.logger.Errorf("Failed registering block epoch notifier")
+		return
+	}
+
+	defer blockNotifier.Cancel()
+
 	for {
 		select {
-		case event, open := <-r.btcClient.BlockEventChan():
+		case epoch, open := <-blockNotifier.Epochs:
 			if !open {
 				r.logger.Errorf("Block event channel is closed")
 				return // channel closed
 			}
 
+			tip := r.btcCache.Tip()
+
+			// Determine if a reorg happened to we know which flow to continue
+			// if the new block has the same height but a different hash.
+			reorg := false
+			if tip != nil {
+				if epoch.Height < tip.Height ||
+					(epoch.Height == tip.Height && epoch.BlockHeader.BlockHash() != tip.Header.BlockHash()) {
+					reorg = true
+				}
+			}
+
 			var errorRequiringBootstrap error
-			if event.EventType == types.BlockConnected {
-				errorRequiringBootstrap = r.handleConnectedBlocks(event)
-			} else if event.EventType == types.BlockDisconnected {
-				errorRequiringBootstrap = r.handleDisconnectedBlocks(event)
+			if !reorg {
+				errorRequiringBootstrap = r.handleConnectedBlocks(epoch.Height, epoch.BlockHeader)
+			} else {
+				errorRequiringBootstrap = r.handleDisconnectedBlocks(epoch.BlockHeader)
```

Comment:
It seems that we do the same chaining check (from L35 to L45) in both `handleDisconnectedBlocks` and `handleDisconnectedBlocks`. I believe we can combine the logic in one `handleNewBlock()` method. Basically, if the new block cannot connect to the cache tip, we bootstrap. Otherwise, we add it to the cache, no needing of separate logic

Reply:
oh that this good idea 👍 

My initial thought was that we could, translate new block  events into connected / disconnected  block events i.e
If we have the chain:
`H1 -> H2 -> H3 `, now the the reorg happens and the chain becomes `H1 - H2' - H3'`, and now we receive `H3'`  then we can:
1. Find the common ancestor between forks i.e `H1` 
2. Calculate that `H2` and `H3` were removed, so send to disconnection events starting from the top
3. Now send connection events for `H2` and `H3` 

Bootstrapping after reorg is a bit brutal i.e it is costly operation, but given that re-orgs are pretty rare imo it is possible route to take and if we encounter some problems with it we can always go for more fancy approach.